### ISSUE TITLE
Linux AArch64 RuntimeDyld relocation overflows (#8567 specific to Linux only)

### ISSUE THREAD

**Issue by gmarkall** [numba/numba#9001]
I'm breaking out this issue from #8567 to avoid spamming everyone who participated in that issue with notifications as I add notes and debugging info. In summary, on Linux AArch64,

<details>
<summary>test_standalone.py</summary>

```python
import numpy as np
from nvidia.dali import pipeline_def
import nvidia.dali as dali
import nvidia.dali.fn as fn
import nvidia.dali.types as dali_types
from nvidia.dali.plugin.numba.fn.experimental import numba_function


def set_all_values_to_255_batch(out0, in0):
    out0[0][:] = 255


def set_all_values_to_255_sample(out0, in0):
    out0[:] = 255


def set_all_values_to_float_batch(out0, in0):
    out0[0][:] = 0.5


def set_all_values_to_float_sample(out0, in0):
    out0[:] = 0.5


def setup_change_out_shape(out_shape, in_shape):
    out0_shape = out_shape[0]
    in0_shape = in_shape[0]
    perm = [1, 2, 0]
    for sample_idx in range(len(out0_shape)):
        for d in range(len(perm)):
            out0_shape[sample_idx][d] = in0_shape[sample_idx][perm[d]]


def change_out_shape_batch(out0, in0):
    for sample_id in range(len(out0)):
        out0[sample_id][:] = 42


def change_out_shape_sample(out0, in0):
    out0[:] = 42


def get_data(shapes, dtype):
    return [np.empty(shape, dtype=dtype) for shape in shapes]


def get_data_zeros(shapes, dtype):
    return [np.zeros(shape, dtype=dtype) for shape in shapes]

@pipeline_def
def numba_func_pipe(shapes, dtype, run_fn=None, out_types=None, in_types=None,
                    outs_ndim=None, ins_ndim=None, setup_fn=None, batch_processing=None):
    data = fn.external_source(lambda: get_data(shapes, dtype), batch=True, device="cpu")
    return numba_function(
        data, run_fn=run_fn, out_types=out_types, in_types=in_types,
        outs_ndim=outs_ndim, ins_ndim=ins_ndim, setup_fn=setup_fn,
        batch_processing=batch_processing)


def _testimpl_numba_func(shapes, dtype, run_fn, out_types, in_types,
                         outs_ndim, ins_ndim, setup_fn, batch_processing, expected_out):
    batch_size = len(shapes)
    pipe = numba_func_pipe(
        batch_size=batch_size, num_threads=1, device_id=0,
        shapes=shapes, dtype=dtype,
        run_fn=run_fn, setup_fn=setup_fn, out_types=out_types,
        in_types=in_types, outs_ndim=outs_ndim, ins_ndim=ins_ndim,
        batch_processing=batch_processing)
    pipe.build()
    for _ in range(3):
        outs = pipe.run()
        for i in range(batch_size):
            out_arr = np.array(outs[0][i])
            assert np.array_equal(out_arr, expected_out[i])

def test_numba_func():
    # shape, dtype, run_fn, out_types,
    # in_types, out_ndim, in_ndim, setup_fn, batch_processing,
    # expected_out
    args = [
        ([(10, 10, 10)], np.uint8, set_all_values_to_255_batch, [dali_types.UINT8],
         [dali_types.UINT8], [3], [3], None, True,
         [np.full((10, 10, 10), 255, dtype=np.uint8)]),
        ([(10, 10, 10)], np.uint8, set_all_values_to_255_sample, [dali_types.UINT8],
         [dali_types.UINT8], [3], [3], None, None,
         [np.full((10, 10, 10), 255, dtype=np.uint8)]),
        ([(10, 10, 10)], np.float32, set_all_values_to_float_batch, [dali_types.FLOAT],
         [dali_types.FLOAT], [3], [3], None, True,
         [np.full((10, 10, 10), 0.5, dtype=np.float32)]),
        ([(10, 10, 10)], np.float32, set_all_values_to_float_sample, [dali_types.FLOAT],
         [dali_types.FLOAT], [3], [3], None, None,
         [np.full((10, 10, 10), 0.5, dtype=np.float32)]),
        ([(10, 20, 30), (20, 10, 30)], np.int64, change_out_shape_batch, [dali_types.INT64],
         [dali_types.INT64], [3], [3], setup_change_out_shape, True,
         [np.full((20, 30, 10), 42, dtype=np.int32),
          np.full((10, 30, 20), 42, dtype=np.int32)]),
        ([(10, 20, 30), (20, 10, 30)], np.int64, change_out_shape_sample, [dali_types.INT64],
         [dali_types.INT64], [3], [3], setup_change_out_shape, None,
         [np.full((20, 30, 10), 42, dtype=np.int32),
          np.full((10, 30, 20), 42, dtype=np.int32)]),
    ]

    for shape, dtype, run_fn, out_types, in_types, outs_ndim, ins_ndim, \
            setup_fn, batch_processing, expected_out in args:
        _testimpl_numba_func(
            shape, dtype, run_fn, out_types, in_types, outs_ndim, ins_ndim, \
            setup_fn, batch_processing, expected_out
        )

test_numba_func()
```

</details>

results in:

```
$ python test_standalone.py 
python: /root/miniconda3/envs/buildenv/conda-bld/llvmdev_1680642098205/work/llvm/lib/ExecutionEngine/RuntimeDyld/RuntimeDyldELF.cpp:507: void llvm::RuntimeDyldELF::resolveAArch64Relocation(const llvm::SectionEntry&, uint64_t, uint64_t, uint32_t, int64_t): Assertion `isInt<33>(Result) && "overflow check failed for relocation"' failed.
Aborted (core dumped)
```

**Comment by gmarkall** [comment]
It turns out I can even hit the assertion with a smaller test than the original reproducer from #8567 on Linux, in a docker container:

```
$ NUMBA_OPT=0 python -m unittest -vb numba.tests.test_stencils.TestManyStencils.test_basic40 

test_basic40 (numba.tests.test_stencils.TestManyStencils)
2 args! ... python: /root/miniconda3/envs/buildenv/conda-bld/llvmdev_1680642098205/work/llvm/lib/ExecutionEngine/RuntimeDyld/RuntimeDyldELF.cpp:507: void llvm::RuntimeDyldELF::resolveAArch64Relocation(const llvm::SectionEntry&, uint64_t, uint64_t, uint32_t, int64_t): Assertion `isInt<33>(Result) && "overflow check failed for relocation"' failed.
Fatal Python error: Aborted

Current thread 0x0000ffffb74d4010 (most recent call first):
  File "/usr/local/lib/python3.8/dist-packages/llvmlite/binding/ffi.py", line 152 in __call__
  File "/usr/local/lib/python3.8/dist-packages/llvmlite/binding/executionengine.py", line 92 in finalize_object
  File "/usr/local/lib/python3.8/dist-packages/numba/core/codegen.py", line 1060 in wrapper
  File "/usr/local/lib/python3.8/dist-packages/numba/core/codegen.py", line 999 in _finalize_specific
  File "/usr/local/lib/python3.8/dist-packages/numba/core/codegen.py", line 797 in _finalize_final_module
  File "/usr/local/lib/python3.8/dist-packages/numba/core/codegen.py", line 765 in finalize
  File "/usr/local/lib/python3.8/dist-packages/numba/core/codegen.py", line 567 in _ensure_finalized
  File "/usr/local/lib/python3.8/dist-packages/numba/core/codegen.py", line 989 in get_pointer_to_function
  File "/usr/local/lib/python3.8/dist-packages/numba/core/cpu.py", line 236 in get_executable
  File "/usr/local/lib/python3.8/dist-packages/numba/core/typed_passes.py", line 495 in run_pass
  File "/usr/local/lib/python3.8/dist-packages/numba/core/compiler_machinery.py", line 273 in check
  File "/usr/local/lib/python3.8/dist-packages/numba/core/compiler_machinery.py", line 311 in _runPass
  File "/usr/local/lib/python3.8/dist-packages/numba/core/compiler_lock.py", line 35 in _acquire_compile_lock
  File "/usr/local/lib/python3.8/dist-packages/numba/core/compiler_machinery.py", line 356 in run
  File "/usr/local/lib/python3.8/dist-packages/numba/core/compiler.py", line 494 in _compile_core
  File "/usr/local/lib/python3.8/dist-packages/numba/core/compiler.py", line 528 in _compile_bytecode
  File "/usr/local/lib/python3.8/dist-packages/numba/core/compiler.py", line 460 in compile_extra
  File "/usr/local/lib/python3.8/dist-packages/numba/core/compiler.py", line 742 in compile_extra
  File "/usr/local/lib/python3.8/dist-packages/numba/tests/test_stencils.py", line 96 in _compile_this
  File "/usr/local/lib/python3.8/dist-packages/numba/tests/test_stencils.py", line 105 in compile_parallel
  File "/usr/local/lib/python3.8/dist-packages/numba/tests/test_stencils.py", line 749 in check_against_expected
  File "/usr/local/lib/python3.8/dist-packages/numba/tests/test_stencils.py", line 1844 in test_basic40
  File "/usr/lib/python3.8/unittest/case.py", line 633 in _callTestMethod
  File "/usr/lib/python3.8/unittest/case.py", line 676 in run
  File "/usr/lib/python3.8/unittest/case.py", line 736 in __call__
  File "/usr/lib/python3.8/unittest/suite.py", line 122 in run
  File "/usr/lib/python3.8/unittest/suite.py", line 84 in __call__
  File "/usr/lib/python3.8/unittest/suite.py", line 122 in run
  File "/usr/lib/python3.8/unittest/suite.py", line 84 in __call__
  File "/usr/lib/python3.8/unittest/runner.py", line 176 in run
  File "/usr/lib/python3.8/unittest/main.py", line 271 in runTests
  File "/usr/lib/python3.8/unittest/main.py", line 101 in __init__
  File "/usr/lib/python3.8/unittest/__main__.py", line 18 in <module>
  File "/usr/lib/python3.8/runpy.py", line 87 in _run_code
  File "/usr/lib/python3.8/runpy.py", line 194 in _run_module_as_main
```

**Comment by gmarkall** [comment]
I now have a simpler reproducer, which works with Numba and llvmlite `main` built on my Jetson AGX Xavier against the Numba `llvmdev` package, running on a "plain" Ubuntu 20.04 (i.e. no docker, no additional tools / libraries needed):

```python
from numba import njit

@njit
def f(x, y):
    return x + y

i = 0

while True:
    print(i)
    t = tuple(range(i))
    f(t, (1j,))
    i += 1
```

executed with:

```
$ ulimit -s 1048576
$ python repro.py
```

gives:

```
0
1
2
3
4
5
6
7
8
9
python: /opt/conda/conda-bld/llvmdev_1684517249134/work/llvm/lib/ExecutionEngine/RuntimeDyld/RuntimeDyldELF.cpp:507: void llvm::RuntimeDyldELF::resolveAArch64Relocation(const llvm::SectionEntry&, uint64_t, uint64_t, uint32_t, int64_t): Assertion `isInt<33>(Result) && "overflow check failed for relocation"' failed.
Aborted (core dumped)
```

**Comment by gmarkall** [comment]
Repository for reproducer work: https://github.com/gmarkall/numba-issue-9001

I now have a repro that only uses NumPy and llvmlite, not Numba, which is the present state recorded in that repo.

**Comment by zansibal** [comment]
Hi, I don't know if this is helpful, but I thought I'd share my experience of this issue. I have come across the issue with my code both on the Ampere 80 core machine running Ubuntu and on AWS Graviton arm64 instances running Amazon Linux 2. On smaller AWS instances (like 8 cores) the problem is less frequent than on larger instances (like 32/64 cores or the Ampere 80). Perhaps this is because smaller instances have less RAM (8 GB vs 128 GB)?

I can reproduce the numba-issue-9001 code on AWS m7gd.2xlarge after only 5 iterations.

**Comment by gmarkall** [comment]
Thanks for the input - when there's a lot of RAM it does seem that it's easier for the issue to manifest in our experience too.

There's also a start of a discussion on the LLVM Discourse regarding a potential fix: https://discourse.llvm.org/t/llvm-rtdyld-aarch64-abi-relocation-restrictions/74616

**Comment by gmarkall** [comment]
@zansibal It's early work at the moment, but if you're able to build llvmlite from source with the PR numba/llvmlite#1009, and let me know whether you still observe the issue with it (or observe any other issues) that would be good feedback - hopefully this resolves the issue, but there's a lot of testing / review to be done to have confidence in the strategy.

**Comment by gmarkall** [comment]
This was fixed by numba/llvmlite#1009.
